src/app/clusters/bindings/bindings.cpp

```diff
@@ -20,155 +20,265 @@
-    EmberBindingTableEntry bindingEntry;
-
-    ChipLogDetail(Zcl, "RX: UnbindCallback");
-
-    if ((groupId != 0 && nodeId != 0) || (groupId == 0 && nodeId == 0))
-    {
-        ChipLogError(Zcl, "Binding: Invalid request");
-        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_MALFORMED_COMMAND);
-        return true;
-    }
-    if (groupId)
-    {
-        bindingEntry = EmberBindingTableEntry::ForGroup(fabricIndex, groupId, localEndpoint, clusterId);
-    }
-    else
+    switch (path.mAttributeId)
     {
-        bindingEntry = EmberBindingTableEntry::ForNode(fabricIndex, nodeId, localEndpoint, remoteEndpoint, clusterId);
+    case Binding::Attributes::BindingList::Id:
+        return WriteBindingTable(path, decoder);
+    default:
+        break;
     }
+    return CHIP_NO_ERROR;
+}
 
-    uint8_t bindingIndex;
-    if (getBindingIndex(bindingEntry, &bindingIndex) != EMBER_SUCCESS)
+CHIP_ERROR BindingTableAccess::WriteBindingTable(const ConcreteDataAttributePath & path, AttributeValueDecoder & decoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    if (!path.IsListOperation() || path.mListOp == ConcreteDataAttributePath::ListOperation::ReplaceAll)
     {
-        emberAfSendImmediateDefaultResponse(EMBER_ZCL_STATUS_NOT_FOUND);
-        return true;
-    }
+        printf("replace all\n");
+        DecodableBindingListType newBindingList;
+
+        ReturnErrorOnFailure(decoder.Decode(newBindingList));
+        ReturnErrorOnFailure(CheckValidBindingList(newBindingList));
+
+        // Remove entries not in the new binding list
+        for (uint8_t i = 0; i < EMBER_BINDING_TABLE_SIZE; i++)
+        {
+            EmberBindingTableEntry entry;
+            emberGetBinding(i, &entry);
+            if (entry.type != EMBER_UNUSED_BINDING && !IsInBindingList(path.mEndpointId, entry, newBindingList))
+            {
 
-    CHIP_ERROR err = BindingManager::GetInstance().UnicastBindingRemoved(bindingIndex);
-    if (err != CHIP_NO_ERROR)
+                if (entry.type == EMBER_UNICAST_BINDING)
+                {
+                    err = BindingManager::GetInstance().UnicastBindingRemoved(i);
+                    if (err != CHIP_NO_ERROR)
+                    {
+                        ChipLogError(Zcl,
+                                     "Binding: Failed to remove pending notification for unicast binding " ChipLogFormatX64
+                                     ": %" CHIP_ERROR_FORMAT,
+                                     ChipLogValueX64(entry.nodeId), err.Format());
+                    }
+                }
+            }
+        }
+        emberClearBinding();
+
+        // Add entries currently not in the binding table
+        uint8_t bindingIndex = 0;
+        for (auto iter = newBindingList.begin(); iter.Next();)
+        {
+            uint8_t oldIndex = 0;
+            if (iter.GetValue().nodeId.HasValue() && IsInBindingTable(path.mEndpointId, iter.GetValue(), &oldIndex))
```

Comment:
So there are two problems here:
1. We cleared the table above, so this is not going to test what we _wanted_ it to test.
2. If the incoming list has two identical entries in it (which is silly, I agree) we're going to drop one of them.

We need some other algorithm here.  The simplest thing to do would be to walk the incoming list, for each item in the list find the next spot in our table that is either available or has the same fabric index, and write it to that spot.  Then once the list runs out clear out all remaining stale entries for the given fabric index.

The hard part there is not generating spurious "unicast binding removed" notifications, right?

But those have another problem.  In practice, the codepath that will get taken will be the `path.mListOp == ConcreteDataAttributePath::ListOperation::AppendItem` codepath, because the client will chunk up the write....

So the right thing to do is probably just to dispatch the removed/added notifications while doing the simple thing here, and if we need to debounce to avoid thrashing our CASE connections we should probably do that in the code that gets those notifications.  And probably as a separate PR.